Implement `some` elision behind an experimental feature flag

angela-laar opened the pull request:
Plain protocol names can be written to mean `some` instead of `any` with the experimental feature flag: `-enable-experimental-implicit-some`.Meaning the `some` keyword would be optional for opaque parameters and opaque result types while enforcing explicit use of the `any` keyword. 

With `some` as the new default for plain protocols (when using this flag) we can avoid unintentially opting into the limitations that come with `any` being the current default for plain protocols. i.e. the inability to access Self and associated type requirements. 

`func join(_ one: Collection, _ two: Collection) -> Collection  { ... } `

would be equivalent to:

`func join(_ one: some Collection, _ two: some Collection) -> some Collection  { ... } `
 
 
rdar://98153760

angela-laar committed:
Add experimental feature flag for Implement 'some'

angela-laar committed:
[AST] Teach GenericParamList amd OpaqueReturnType Request to use plain protocols

Originally protocols prefixed with 'some' keyword are considered an opaue return
type representation. In order to understand plain protocols, the GenericParamList
Request must understand how to build a generic parameter list using alternative
type representations: identifier and composition.

angela-laar committed:
[AST] Namelookup should only look at parsed generic parameters

We want to make sure not to cause a circular reference
by fetching the full generic parameter list when we are
creating opaque paramters and result types.

angela-laar committed:
[AST] Verify identifier type representations are protocols

Now that GenericParamListRequest and OpaqueResultTypeRequest know
to check for IdentTypeReprs when scanning for opaque parameters,
the requests need to know it should only create Generic Paramters
and Opaque Type Decls for IdentTypeReprs that are protocols.
We should add a function to the swift namespace to check.

angela-laar committed:
Avoid walking the nested type reprs for types reprs marked with 'some' keyword

angela-laar committed:
Prevent existential_requires_any error

angela-laar committed:
[Sema] Create reference to id reprs for type resolution

angela-laar committed:
[AST] Preserve type constraint for identifier type

Type contraints need to be passed when creating
generic parameters from identifier types.

angela-laar committed:
[AST] Ignore existential types when building opaque parameter list

angela-laar commented:
@swift-ci please test

angela-laar commented:
@swift-ci please test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

hborla commented on include/swift/AST/NameLookup.h:
```diff
@@ -27,21 +27,38 @@
+CollectedOpaqueReprs collectTypeReprs(TypeRepr *);
+
+/// Given a Type Representation, decide if it is a protocol
+bool isProtocol(Evaluator &evaluator, TypeRepr *r, ASTContext &ctx, DeclContext *dc);
```
Nit: `ASTContext` has an `evaluator` property, so you can access the evaluator through `ctx.evaluator` rather than passing it separately!

Also, should this be a method on `TypeRepr` instead of a top-level function?

angela-laar replied:
`isProtocol` does make more sense as a TypeRepr method, updated!

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2780,12 +2794,58 @@ createExtensionGenericParams(ASTContext &ctx,
 static SmallVector<GenericTypeParamDecl *, 2>
-createOpaqueParameterGenericParams(
-    GenericContext *genericContext, GenericParamList *parsedGenericParams) {
+createOpaqueParameterGenericParams(Evaluator &evaluator,GenericContext *genericContext, GenericParamList *parsedGenericParams) {
```
Nit: You don't need to pass `evaluator` here either - you can access it through the `ASTContext` from the generic context on the next line

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
+    CollectedOpaqueReprs typeReprs;
+    if (ctx.LangOpts.hasFeature(Feature::ImplicitSome)) {
+        typeReprs = collectTypeReprs(typeRepr);
+    } else {  typeReprs = collectOpaqueReturnTypeReprs(typeRepr); }
```
Should we check for `Feature::ImplicitSome` inside of `collectOpaqueReturnTypeReprs` instead of having two separate functions here? I think that should be pretty straightforward, since we only want to collect `OpaqueReturnTypeRepr` if the experimental feature is not enabled.

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
+      for (auto repr : typeReprs) {
+   
+        if (isa<IdentTypeRepr>(repr)){
+            if(!isProtocol(evaluator,repr, ctx, dc))
```
Should we check for `isProtocol` inside of `collectOpaqueReturnTypeReprs` and only collect the `IdentTypeRepr` if it represents a protocol?

angela-laar replied:
updated!

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
+        }
       // Allocate a new generic parameter to represent this opaque type.
-      auto gp = GenericTypeParamDecl::create(
+        auto gp = GenericTypeParamDecl::create(
```
Nit: Looks like there was an accidental space added here and on line 2892

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
 }
 
+
+bool swift::isProtocol( Evaluator &evaluator, TypeRepr *r, ASTContext &ctx, DeclContext *dc) {
```
Nit: Same comment about the `evaluator` parameter here!

hborla commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
+
+bool swift::isProtocol( Evaluator &evaluator, TypeRepr *r, ASTContext &ctx, DeclContext *dc) {
+    DirectlyReferencedTypeDecls d = directReferencesForTypeRepr(evaluator, ctx, r, dc);
+    if(declsAreProtocols(d)){
```
Since this is returning true if `declsAreProtocols(d)` is true and false otherwise, this can be simplified to just `return declsAreProtocols(d)`

hborla commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -134,34 +136,45 @@ OpaqueResultTypeRequest::evaluate(Evaluator &evaluator,
       auto *currentRepr = opaqueReprs[i];
+        
+      if (isa<IdentTypeRepr>(currentRepr)){
+        if(!isProtocol(evaluator, currentRepr, ctx, dc))
```
You mentioned this to me offline but I'll record it here 🙂 - if we check `isProtocol` inside of `collectOpaqueReturnTypeReprs`, we won't need to duplicate this code here and inside `createOpaqueParameterGenericParams`.

angela-laar replied:
yes! updated 💯

hborla commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -134,34 +136,45 @@ OpaqueResultTypeRequest::evaluate(Evaluator &evaluator,
+                              currentRepr->getEndLoc(), stream.str());
         return nullptr;
-      }
+      }}
```
Nit: Add the new closing brace on the next line. We should also indent the code inside your new condition (from 159-171) two spaces.

hborla commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -189,7 +202,7 @@ OpaqueResultTypeRequest::evaluate(Evaluator &evaluator,
 
       assert(!constraintType->hasArchetype());
-      requirements.emplace_back(kind, paramType, constraintType);
+      requirements.emplace_back(kind, paramType, constraintType);}
```
Hmm, I don't think we want this bit of code inside an `OpaqueReturnTypeRepr` condition. We want to add a new requirement for implicit opaque return type reps too. The only difference is the `constraintType` will be resolved directly from the collected `IdentTypeRepr` or a `CompositionTypeRepr` instead of the constraint type representation of an `OpaqueReturnTypeRepr`.

hborla reviewed (commented):
Awesome progress! I've left a bunch of comments inline, mostly minor stylistic comments 🙂

slavapestov commented:
Also, should we rename OpaqueReturnTypeRepr to OpaqueTypeRepr at some point?

slavapestov commented on include/swift/AST/ASTWalker.h:
```diff
@@ -20,14 +20,17 @@
 namespace swift {
 
 class ArgumentList;
+class ASTContext;
```
Are these changes to ASTWalker.h still needed, since there were no other changes in this file?

slavapestov commented on include/swift/AST/NameLookup.h:
```diff
@@ -27,21 +27,38 @@
+
+/// Walk the type representation recursively, collecting any
+/// `CompositionTypeRepr`s and `IdentTypeRepr`s.
+CollectedOpaqueReprs collectTypeReprs(TypeRepr *);
```
This name is too general since it doesn't collect _all_ type representations, only the protocol (and protocol composition presumably). How about collectProtocolTypeReprs instead?

angela-laar replied:
I renamed this to `collectOpaqueReturnTypeReprs` and removed the second function in my latest commit. 
Is it too soon to just call this `collectOpaqueTypeReprs` ? 👀

slavapestov commented on include/swift/AST/NameLookup.h:
```diff
@@ -27,21 +27,38 @@
+CollectedOpaqueReprs collectTypeReprs(TypeRepr *);
+
+/// Given a Type Representation, decide if it is a protocol
+bool isProtocol(Evaluator &evaluator, TypeRepr *r, ASTContext &ctx, DeclContext *dc);
```
In addition to getting the evaluator from the ASTContext, you can also get the ASTContext from the DeclContext by calling getASTContext()

slavapestov commented on include/swift/AST/NameLookup.h:
```diff
@@ -27,21 +27,38 @@
 class ASTScopeImpl;
 } // namespace ast_scope
 
+using CollectedOpaqueReprs = SmallVector<TypeRepr *, 2>;
```
Since NameLookup.h already includes TypeRepr.h, you don't need to restate the type alias here

slavapestov commented on include/swift/AST/TypeRepr.h:
```diff
@@ -843,7 +840,7 @@ class TupleTypeRepr final : public TypeRepr,
 /// \endcode
 class CompositionTypeRepr final : public TypeRepr,
-    private llvm::TrailingObjects<CompositionTypeRepr, TypeRepr*> {
+  private llvm::TrailingObjects<CompositionTypeRepr, TypeRepr*> {
```
Indentation nit: we use 4 spaces here

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -2780,12 +2794,58 @@ createExtensionGenericParams(ASTContext &ctx,
+      } else if (auto existential = dyn_cast<ExistentialTypeRepr>(repr)) {
+          return false;
+      } else if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+          Reprs.push_back(identRepr);
```
Maybe the isProtocol() check can go here?

angela-laar replied:
updated!

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -2812,28 +2872,53 @@ createOpaqueParameterGenericParams(
-      gp->setInherited(ctx.AllocateCopy(inherited));
-
+      //  The underlying constraint is only present for OpaqueReturnTypeReprs
+        if( auto opaque = dyn_cast<OpaqueReturnTypeRepr>(repr)){
```
There's an extra level of indent here, and you should transpose the ` ` and `(`

slavapestov commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -690,9 +690,12 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
     Opts.Features.insert(Feature::OpaqueTypeErasure);
+  if (Args.hasArg(OPT_enable_experimental_implicit_some)){
+      Opts.Features.insert(Feature::ImplicitSome);
+      Opts.Features.insert(Feature::ExistentialAny);
```
Isn't ExistentialAny on by default?

angela-laar replied:
only for swift 6 mode!

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3730,6 +3730,24 @@ TypeResolver::resolveIdentifierType(IdentTypeRepr *IdType,
+    auto *dc = getDeclContext();
+    auto &ctx = getASTContext();
+
+    if ( ctx.LangOpts.hasFeature(Feature::ImplicitSome) & !options.is(TypeResolverContext::Inherited) ){
```
Perhaps you want to check options.isConstraintImplicitExistential() instead of options.is(TypeResolverContext::Inherited)?

slavapestov commented on test/type/implicit_some/opaque_parameters.swift:
```diff
@@ -0,0 +1,108 @@
+// RUN: %target-typecheck-verify-swift -disable-availability-checking -warn-redundant-requirements  -enable-experimental-implicit-some
```
You should test type aliases where the underlying type is a protocol type or composition, as well

slavapestov commented on test/type/implicit_some/opaque_parameters.swift:
```diff
@@ -0,0 +1,108 @@
+}
+
+// Combine with parameterized protocol types
+protocol PrimaryCollection<Element>: Collection {}
```
Probably this protocol isn't needed anymore because Collection has a primary associated type now

slavapestov reviewed (commented):
Looks great!

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Resolve ident type to opaque result type

This is just more work to get type resolution to understand
how to derive an opaque type from an IdentTypeRepr during
OpaqueResultTypeRequest

angela-laar committed:
[Sema] Support GenericIdentTypeReprs as opaque parameters and return type

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[AST] Create opaque type from tuple type repr

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[AST] Add new ASTWalker actions to opaque type repr walker

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Create Opaque Decls from Composition Types

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[test] fix tests

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci Please smoke test macOS platform

angela-laar commented on lib/AST/NameLookup.cpp:
```diff
@@ -2783,12 +2800,65 @@ createExtensionGenericParams(ASTContext &ctx,
+      
+      if (auto existential = dyn_cast<ExistentialTypeRepr>(repr)) {
+        auto meta = dyn_cast<MetatypeTypeRepr>(existential->getConstraint());
+        auto generic = dyn_cast<GenericIdentTypeRepr>(existential->getConstraint());
```
We need to check if `any P<>` is wrapping an opaque generic or meta type
